### DISCUSSION TITLE
ST7789 cannot get it to work

### DISCUSSION THREAD

**Discussion by phhoef** [lovyan03/LovyanGFX#578] [Q&A]
## Environment ( 実行環境 )

- MCU or Board name:  ESP32-S3-WROOM-1-N16R8
- Panel Driver IC:    ST7789
- Bus type:           SPI
- LovyanGFX version:  v1.1.16
- FrameWork version:   ArduinoESP32 v2.0.3
- Build Environment:   PlatformIO
- Operating System:   macOS

## Problem Description ( 問題の内容 )
I am using TFT_eSPI with LVGL currently. This is working, so I am pretty sure the wiring is correct.
Now, I would like to switch to LovyanGFX and I cannot get it work.

I've removed the touch and light_pwm part.
I do not have touch controller, that's why I left out this passage.
The backlight is also burning, but cannot be controlled through pwm. Though it can be turned on and off. That's why I removed the light_pwm passage.

The display is just black and nothing happens.
ESP32 is not crashing.

### Code to reproduce this issue ( 再現させるためのコード )

**Please submit complete source code that can reproduce your problem.**
**あなたの問題を再現できる完全なソースコードを提示してください。**

```cpp
#include <LovyanGFX.hpp>

class LGVX : public lgfx::LGFX_Device
{
    lgfx::Panel_ST7789 _panel_instance;

    // パネルを接続するバスの種類にあったインスタンスを用意します。
    lgfx::Bus_SPI _bus_instance; // SPIバスのインスタンス
                                 // lgfx::Bus_I2C        _bus_instance;   // I2Cバスのインスタンス
    // lgfx::Bus_Parallel8  _bus_instance;   // 8ビットパラレルバスのインスタンス

public:
    // コンストラクタを作成し、ここで各種設定を行います。
    // クラス名を変更した場合はコンストラクタも同じ名前を指定してください。
    LGVX(void)
    {
        {                                      // バス制御の設定を行います。
            auto cfg = _bus_instance.config(); // バス設定用の構造体を取得します。

            // SPIバスの設定
            cfg.spi_host = SPI3_HOST; // 使用するSPIを選択  ESP32-S2,C3 : SPI2_HOST or SPI3_HOST / ESP32 : VSPI_HOST or HSPI_HOST
            // ※ ESP-IDFバージョンアップに伴い、VSPI_HOST , HSPI_HOSTの記述は非推奨になるため、エラーが出る場合は代わりにSPI2_HOST , SPI3_HOSTを使用してください。
            cfg.spi_mode = 0;                  // SPI通信モードを設定 (0 ~ 3)
            cfg.freq_write = 40000000;         // 送信時のSPIクロック (最大80MHz, 80MHzを整数で割った値に丸められます)
            cfg.freq_read = 16000000;          // 受信時のSPIクロック
            cfg.spi_3wire = true;              // 受信をMOSIピンで行う場合はtrueを設定
            cfg.use_lock = true;               // トランザクションロックを使用する場合はtrueを設定
            cfg.dma_channel = SPI_DMA_CH_AUTO; // 使用するDMAチャンネルを設定 (0=DMA不使用 / 1=1ch / 2=ch / SPI_DMA_CH_AUTO=自動設定)
            // ※ ESP-IDFバージョンアップに伴い、DMAチャンネルはSPI_DMA_CH_AUTO(自動設定)が推奨になりました。1ch,2chの指定は非推奨になります。
            cfg.pin_sclk = 36; // SPIのSCLKピン番号を設定
            cfg.pin_mosi = 35; // SPIのMOSIピン番号を設定
            cfg.pin_miso = -1; // SPIのMISOピン番号を設定 (-1 = disable)
            cfg.pin_dc = 21;   // SPIのD/Cピン番号を設定  (-1 = disable)
                               // SDカードと共通のSPIバスを使う場合、MISOは省略せず必ず設定してください。
                               //*/

            _bus_instance.config(cfg);              // 設定値をバスに反映します。
            _panel_instance.setBus(&_bus_instance); // バスをパネルにセットします。
        }

        {                                        // 表示パネル制御の設定を行います。
            auto cfg = _panel_instance.config(); // 表示パネル設定用の構造体を取得します。

            cfg.pin_cs = 19;   // CSが接続されているピン番号   (-1 = disable)
            cfg.pin_rst = 20;  // RSTが接続されているピン番号  (-1 = disable)
            cfg.pin_busy = -1; // BUSYが接続されているピン番号 (-1 = disable)

            // ※ 以下の設定値はパネル毎に一般的な初期値が設定されていますので、不明な項目はコメントアウトして試してみてください。

            cfg.panel_width = 240;    // 実際に表示可能な幅
            cfg.panel_height = 320;   // 実際に表示可能な高さ
            cfg.offset_x = 0;         // パネルのX方向オフセット量
            cfg.offset_y = 0;         // パネルのY方向オフセット量
            cfg.offset_rotation = 0;  // 回転方向の値のオフセット 0~7 (4~7は上下反転)
            cfg.dummy_read_pixel = 8; // ピクセル読出し前のダミーリードのビット数
            cfg.dummy_read_bits = 1;  // ピクセル以外のデータ読出し前のダミーリードのビット数
            cfg.readable = true;      // データ読出しが可能な場合 trueに設定
            cfg.invert = false;       // パネルの明暗が反転してしまう場合 trueに設定
            cfg.rgb_order = false;    // パネルの赤と青が入れ替わってしまう場合 trueに設定
            cfg.dlen_16bit = false;   // 16bitパラレルやSPIでデータ長を16bit単位で送信するパネルの場合 trueに設定
            cfg.bus_shared = true;    // SDカードとバスを共有している場合 trueに設定(drawJpgFile等でバス制御を行います)

            // 以下はST7735やILI9163のようにピクセル数が可変のドライバで表示がずれる場合にのみ設定してください。
            //    cfg.memory_width     =   240;  // ドライバICがサポートしている最大の幅
            //    cfg.memory_height    =   320;  // ドライバICがサポートしている最大の高さ

            _panel_instance.config(cfg);
        }

        setPanel(&_panel_instance); // 使用するパネルをセットします。
    }
};

void setup()
{
  LGVX *tft = new LGVX();
  tft->init();
  tft->setRotation(1);
  tft->initDMA();

  tft->clear();
  tft->fillRect(30, 30, 100, 100, tft->color565(255, 0, 0));
}
```

This is the `Setup_User.h` from TFT_eSPI.
MISO is not connected to the display.

```cpp

#define USER_SETUP_INFO "User_Setup"

#define ST7789_DRIVER // Full configuration option, define additional parameters below for this display

#define TFT_WIDTH 240 // ST7789 240 x 240 and 240 x 320
#define TFT_HEIGHT 320 // ST7789 240 x 320

#define TFT_BL 11             // LED back-light control pin
#define TFT_BACKLIGHT_ON HIGH // Level to turn ON back-light (HIGH or LOW)

// #define TFT_MISO 37
#define TFT_MOSI 35
#define TFT_SCLK 36
#define TFT_CS 19  // Chip select control pin
#define TFT_DC 21  // Data Command control pin
#define TFT_RST 20 // Reset pin (could connect to RST pin)
// #define TFT_RST  -1  // Set TFT_RST to -1 if display RESET is connected to ESP32 board RST

#define LOAD_GLCD  // Font 1. Original Adafruit 8 pixel font needs ~1820 bytes in FLASH
#define LOAD_FONT2 // Font 2. Small 16 pixel high font, needs ~3534 bytes in FLASH, 96 characters
#define LOAD_FONT4 // Font 4. Medium 26 pixel high font, needs ~5848 bytes in FLASH, 96 characters
#define LOAD_FONT6 // Font 6. Large 48 pixel font, needs ~2666 bytes in FLASH, only characters 1234567890:-.apm
#define LOAD_FONT7 // Font 7. 7 segment 48 pixel font, needs ~2438 bytes in FLASH, only characters 1234567890:-.
#define LOAD_FONT8 // Font 8. Large 75 pixel font needs ~3256 bytes in FLASH, only characters 1234567890:-.
// #define LOAD_FONT8N // Font 8. Alternative to Font 8 above, slightly narrower, so 3 digits fit a 160 pixel TFT
#define LOAD_GFXFF // FreeFonts. Include access to the 48 Adafruit_GFX free fonts FF1 to FF48 and custom fonts

#define SMOOTH_FONT

#define SPI_FREQUENCY 80000000

#define SPI_READ_FREQUENCY 20000000

#define SPI_TOUCH_FREQUENCY 2500000

```

**Comment by lovyan03** [comment]
`cfg.pin_dc = -1;`

why???

**Reply by phhoef** [reply]
true. Sorry, that was a copy&paste mistake. I've updated the initial post.

**Comment by phhoef** [comment]
This is driving me nuts.
I've double checked the pins multiple times and I don't see a mistake.
I tried disabling DMA, but I've always got the same result.
![IMG_1706](https://github.com/lovyan03/LovyanGFX/assets/32458104/6b7f6030-1cd9-49b9-82d9-11d02e92004a)
(The blue/black line across the display is on the protective foil)

Can we narrow the problem down to either the bus or the panel?

**Reply by tobozo** [reply]
you can try to add this to your sketch:

    Serial.begin(115200);

... and set output debug level to "debug" in the platformio.ini file

    build_flags = -DCORE_DEBUG_LEVEL=4

then if a bus error occurs, it should be printed to the platformio monitor

**Reply by phhoef** [reply]
Thanks, I followed your suggestion and got the following output.

```
ESP-ROM:esp32s3-20210327
Build:Mar 27 2021
rst:0x1 (POWERON),boot:0xa (SPI_FAST_FLASH_BOOT)
SPIWP:0xee
mode:DIO, clock div:1
load:0x3fce3808,len:0x44c
load:0x403c9700,len:0xbd8
load:0x403cc700,len:0x2a80
entry 0x403c98d0
[   127][I][esp32-hal-i2c.c:75] i2cInit(): Initialising I2C Master: sda=8 scl=9 freq=100000
[   132][W][Wire.cpp:301] begin(): Bus already started in Master Mode.
[   139][W][Wire.cpp:301] begin(): Bus already started in Master Mode.
```

I2C is used for accelerometer. Would you expect to that his influences the spi bus?
To me it looks unsuspicious.

**Reply by tobozo** [reply]
yup not really suspicious

but anything not related to the LFGX configuration should be disabled until you get a working configuration, including whatever workaround you're using for the backlight

some random thoughts:

- the first instance you should get to work is Light_PWM
- try to use a static LGVX reference from outside the setup() function instead of a scoped pointer e.g. `static LGVX tft;` 
- comment out the line with `tft.initDMA();`, it is not necessary for the configuration tests
- play with the `cfg.spi_host` value
- manually set the pin direction for RST `pinMode( GPIO_NUM_20, OUTPUT )` before calling `tft.init()`

**Reply by phhoef** [reply]
Thanks for your help.

I managed to make Light_PWM work. Due to my circuit I doubt that I can actually dim the light, but it is now controlled with the library. I've an additional transistor besides the led driver to open or close the circuit.
This is due to legacy purpose and probably no longer needed.

I setup a brand new project without any other dependency.
The result is unfortunately the same. I also tried playing with the `spi_host` - no changes. I noticed, that in my main project I cannot use `SPI1_HOST` (esp crashes), but in the test project it is possible.
Defining the `pinMode` has no influence, too.

Displays looks always like shown on the picture.
The light is burning, but the content is black.

This is my test project.
https://github.com/phhoef/lvgx_test

Do you see something suspicious?

**Reply by phhoef** [reply]
I've added TFT_eSPI as comparison. This is working as expected.
Do you see any differences that could cause the problem? I am a little bit lost, as I do not understand the problem.

It seems, that nothing is ever written to the display.
I flashed the eSPI version and draw a green rect. This is working.
Then I switched set `USE_LOVYAN=1` and flash again.

The display still shows the green rect. From my point of view this was still in memory of the display.
Would that mean, that the bus is not properly initialised or the bus is never written?

**Reply by tobozo** [reply]
> The display still shows the green rect. From my point of view this was still in memory of the display.

interesting, could this mean the reset isn't performed?

one way to verify is to play with the RST pin from your `USE_LOVYAN` code block, before `tft.init()`; if you flash the eSPI version again, then switch `USE_LOVYAN=1` and flash again, the reset should clear the display:

```cpp
  pinMode( GPIO_NUM_20, OUTPUT );
  digitalWrite( GPIO_NUM_20, 0 ); // also try with 1
  delay(100);
```

**Reply by phhoef** [reply]
I tried manually setting `GPIO_NUM_20` to `0` and `1`. In both cases the content of the display is black, when flashing lovyan version after eSPI.
I put the rule to test, and removed the manual rst and in fact the green rect is still visible after flashing lovyan.

The manual rst seems to work.
But the rst through the library does obviously not. Not sure what is going on ...
Is there a debug flag or something similar to better understand what's going on?
It feels like no matter what we configure, the lib is not initialising properly. I am no expert at all these display stuff, but I assume the RST will be just set, right?

I've committed my code changes.

**Comment by tobozo** [comment]
`cfg.dlen_16bit` should be set to false, the bus isn't 16bits parallel !

    cfg.dlen_16bit = false;

also you can set `cfg.dma_channel` to the default value, it won't hurt

    cfg.dma_channel = SPI_DMA_CH_AUTO

**Reply by phhoef** [reply]
thank you very much! It is working now.
I changed the `dlen_16bit` at some point during my tests and didn't revert it.
That said, I found out that the `RST` needs to be set manually. Otherwise it won't work and display content is always black, as before.
Can you imagine why the `RST` is not set automatically during initialisation?

I'll now try to bring this setup to my main project and make it work with lvgl.

**Reply by tobozo** [reply]
>  Can you imagine why the RST is not set automatically during initialisation?

I've already seen that once on a parallel display with incomplete configuration but it was temporary

However if `digitalWrite( GPIO_NUM_20, 0 );` (LOW) solved it then LovyanGFX is probably to blame as it [pulls the pin HIGH](https://github.com/lovyan03/LovyanGFX/blob/master/src/lgfx/v1/panel/Panel_Device.cpp#L299-L305), otherwise you may remove both pinMode and digitalWrite from your setup and it should work.

**Reply by tobozo** [reply]
btw if you have to keep this pinMode/digitalWrite to get your display working, I suggest you move them inside your LVGX class

```cpp
// (...)
public:
    // コンストラクタを作成し、ここで各種設定を行います。
    // クラス名を変更した場合はコンストラクタも同じ名前を指定してください。
    LGVX(void)
    {
      lgfx::gpio_hi(GPIO_NUM_20);
      lgfx::pinMode(GPIO_NUM_20, pin_mode_t::output);
// (...)
```

and also set `cfg.pin_rst = -1;` to prevent `Panel_Device::init_rst()` to pull it high

**Reply by phhoef** [reply]
Good Morning 😄 

now things are getting strange 😉 

1. I set the `RST` pin to `0` and `1` and in both cases it was working. So it seems, that it is not important which state I am setting.
2. When removing the whole code block, then it stops working. So, it must be related to the `pinMode`.
3. I removed everything expect the `pinMode` and it is still working 

Then I followed your suggestion and moved it to the constructor of my `LGVX` class.
At the very end I inserted
```cpp
lgfx::pinMode(_panel_instance.config().pin_rst, lgfx::output);
```

And it stopped working.
I moved that line around and it never worked. In fact not working means, that the screen becomes black and the old drawn buffer is not visible anymore.
Then I used the `pinMode` function from Arduino framework and it started working again.
It seems, that something different is happening between these functions. I do not understand why this makes a difference.

Updates are pushed to the repo.
Can you make sense of it?

**Reply by tobozo** [reply]
as said, if you trigger the RST pin manually, also set `pin_rst=-1` in the config, or expect strange behaviors

moreover, if pin20 is high when the esp32 boots, it's probably because it's already used by something else such as USB D+, also using it with your display could prevent USB from working

![image](https://github.com/lovyan03/LovyanGFX/assets/1893754/5781c109-0242-4589-98d2-8dc65227a73f)

**Comment by phhoef** [comment]
Small update: It's also working with lvgl now, with the shown configuration.
I assume the example for lvgl was written with v8. The output was not sharp. I am using lvgl v9.
I modified the `flush_cb` method, and basically do it the same way as with eSPI.
```cpp
static void flush_cb(lv_display_t *disp, const lv_area_t *area, uint8_t *px_map)
{
    uint32_t w = (area->x2 - area->x1 + 1);
    uint32_t h = (area->y2 - area->y1 + 1);

    LGVX *tft = (LGVX *)lv_display_get_driver_data(disp);
    if (tft->getStartCount())
        tft->startWrite();

    tft->setAddrWindow(area->x1, area->y1, w, h);
    tft->pushPixelsDMA((uint16_t *)px_map, w * h, true);

    lv_display_flush_ready(disp);
}
``` 

Not sure where exactly the difference between `pushImage` and `pushPixels` is.
I am assuming, that I can pass the position to `pushImage` as function parameter and I do not need to call `setAddrWindow`.

**Comment by vednildip** [comment]
can  you help  to   share  pin  config   as   I  am  having  Lilygo  T  HMI lgfx::Bus_Parallel8  with  MCU or Board name: ESP32-S3-N16R8  and  XPT 2046   having  RD -1 , RET -1 .  

#define PWR_EN_PIN  (10)
#define PWR_ON_PIN  (14)
#define BAT_ADC_PIN (5)
#define BUTTON1_PIN (0)
#define BUTTON2_PIN (21)

// lcd
#define LCD_DATA0_PIN (48)
#define LCD_DATA1_PIN (47)
#define LCD_DATA2_PIN (39)
#define LCD_DATA3_PIN (40)
#define LCD_DATA4_PIN (41)
#define LCD_DATA5_PIN (42)
#define LCD_DATA6_PIN (45)
#define LCD_DATA7_PIN (46)
#define PCLK_PIN      (8)
#define CS_PIN        (6)
#define DC_PIN        (7)
#define RST_PIN       (-1)
#define BK_LIGHT_PIN  (38)

// touch screen
#define TOUCHSCREEN_SCLK_PIN (1)
#define TOUCHSCREEN_MISO_PIN (4)
#define TOUCHSCREEN_MOSI_PIN (3)
#define TOUCHSCREEN_CS_PIN   (2)
#define TOUCHSCREEN_IRQ_PIN  (9)

// sd card
#define SD_MISO_PIN (13)
#define SD_MOSI_PIN (11)
#define SD_SCLK_PIN (12)

#define SDIO_DATA0_PIN (13)
#define SDIO_CMD_PIN   (11)
#define SDIO_SCLK_PIN  (12)

**Reply by phhoef** [reply]
sorry I cannot really help. I am driving my display with SPI. Never used another interface
